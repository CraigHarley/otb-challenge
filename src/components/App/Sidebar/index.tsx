import React, {ReactElement} from 'react';
import {css, StyleSheet} from 'aphrodite';

import {ISort} from '../../../types';
import {SortTypes} from '../../../types/enums';
import config from '../../../config'

export interface ISidebarProps {
    setChosenSort: (sort: SortTypes) => void;
    chosenSort: SortTypes;
    sorts: ISort[];
}

const styles = StyleSheet.create({
    buttonContainer: {
        minWidth: '200px',
        maxWidth: '200px',
        margin: '2rem 5rem'
    },
    button: {
        textTransform: 'lowercase',
        textAlign: 'left',
        appearance: 'none',
        textDecoration: 'none',
        cursor: 'pointer',
        border: 'none',
        height: '3rem',
        minWidth: '100%',
        color: config.styles.colors.primary,
        backgroundColor: 'whitesmoke',
        borderTop: `1px solid ${config.styles.colors.primary}`,
        padding: '0.5rem 1rem 0.5rem 1rem',
        lineHeight: '1rem',
        fontSize: 'small',
        ':hover': {
            backgroundColor: config.styles.colors.secondary,
            color: config.styles.colors.primary,
        }
    },
    firstButton: {
        borderTop: 'none'
    },
    isActive: {
        backgroundColor: config.styles.colors.primary,
        color: 'whitesmoke',
        ':hover': {
            backgroundColor: config.styles.colors.primary,
            color: 'whitesmoke',
        }
    }
});

const Sidebar = ({sorts, chosenSort, setChosenSort}: ISidebarProps): ReactElement => (
    <aside>
        <div className={css(styles.buttonContainer)}>
            {
                sorts.map(
                    ({name, icon, type}, key) => (
                        <button
                            name={`btn-sort-${key}`}
                            key={key}
                            className={
                                css(
                                    styles.button,
                                    (chosenSort === type) && styles.isActive,
                                    key === 0 && styles.firstButton
                                )
                            }
                            onClick={() => setChosenSort(type)}
                        >
                            Sort <strong>{name}</strong> {icon}
                        </button>
                    )
                )
            }
        </div>
    </aside>
);

export default Sidebar;
