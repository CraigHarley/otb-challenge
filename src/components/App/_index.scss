* {
  box-sizing: border-box;
}

body {
  background-image: url("/assets/background.png");
  background-position-y: -300px; // todo relative value
  background-repeat: repeat-x;
  background-color: rgb(251, 252, 250);
  font-family: "Roboto Light", sans-serif;
}

.grid-container {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr;
  gap: 0 0;
  grid-template-areas: "Sidebar Content Content Content";
}

.float-right {
  float: right;
}
