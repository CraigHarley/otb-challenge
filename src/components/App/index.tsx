import React, {ReactElement, useState} from 'react';
import {css, StyleSheet} from "aphrodite";

import {SortTypes} from '../../types/enums';
import {hotels, sorts} from '../../models';
import Sidebar from '../Sidebar';
import Card from '../Card';

const styles = StyleSheet.create({
    gridContainer: {
        display: 'grid',
        gridTemplateColumns: '1fr 1fr 1fr 1fr',
        gridTemplateRows: '1fr',
        gridTemplateAreas: "Sidebar Content Content Content",
    }
})

const App = (): ReactElement => {
    const [chosenSort, setChosenSort] = useState(SortTypes.ALPHA);

    return (
        <div className={css(styles.gridContainer)}>
            <Sidebar
                setChosenSort={setChosenSort}
                chosenSort={chosenSort}
                sorts={sorts}
            />
            <article>
                {
                    hotels
                        .sort(
                            sorts
                                .find((sort) => sort.type === chosenSort)
                                .func
                        )
                        .map((hotel, key) => <Card hotel={hotel} key={key}/>)
                }
            </article>
        </div>
    );
};

export default App;
