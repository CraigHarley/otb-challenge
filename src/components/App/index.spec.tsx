import React from 'react';
import {mount, render} from 'enzyme';

import App from '.';

describe('App', () => {
    test('it should render',
        () =>
            expect(
                render(
                    <App/>
                )
            )
                .toMatchSnapshot('default')
    );

    test('when different sorts are clicked, it should rearrange the hotels',
        () => {
            const element = mount(
                <App/>
            );

            element
                .find('button[name^="btn-sort"]')
                .forEach((btn, key) => {
                    btn
                        .simulate('click')

                    expect(
                        element.render()
                    ).toMatchSnapshot(`sorting: ${key}`);
                })
            ;
        }
    );

})

