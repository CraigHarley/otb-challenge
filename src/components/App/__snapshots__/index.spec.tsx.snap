// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App it should render: default 1`] = `
<div
  class="gridContainer_eg8q4b"
>
  <aside>
    <div
      class="buttonContainer_15ic8la"
    >
      <button
        class="button_uw9t0c-o_O-isActive_1a0itbg-o_O-firstButton_4ebdec"
      >
        Sort 
        <strong>
          alphabetically
        </strong>
         
        <i
          class="float-right fas fa-sort-alpha-down"
        />
      </button>
      <button
        class="button_uw9t0c"
      >
        Sort 
        <strong>
          by price
        </strong>
         
        <i
          class="float-right fas fa-pound-sign"
        />
      </button>
      <button
        class="button_uw9t0c"
      >
        Sort 
        <strong>
          by star rating
        </strong>
         
        <i
          class="float-right fas fa-star"
        />
      </button>
    </div>
  </aside>
  <article>
    <section>
      <div
        class="card_pmicl5"
      >
        <img
          alt="Image of the hotel"
          class="img_bm229d"
          src="http://localhost:9999/hotel-image-1.png"
        />
        <div
          class="contentContainer_1ydb5eu"
        >
          <h1>
            Iberostar Grand Salome
          </h1>
          <h2
            class="h2_1tkgvlr"
          >
            Costa Adeje, Tenerife
          </h2>
          <div>
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
          </div>
          <ul
            class="ul_tx6wom"
          >
            <li>
              <span>
                <strong>
                  2
                </strong>
                 adults & 
                <strong>
                  1
                </strong>
                 infant
              </span>
            </li>
            <li>
              <strong>
                3rd July 2019
              </strong>
               for 
              <strong>
                7 days
              </strong>
            </li>
            <li>
              departing from 
              <strong>
                East Midlands
              </strong>
            </li>
          </ul>
          <button
            class="callToAction_55n7hg"
          >
            <span
              class="actionText_131r5dh"
            >
              Book now
            </span>
            <span
              class="actionAmount_qtg89e"
            >
              £3,136.50
            </span>
          </button>
        </div>
        <button
          class="readMore_j9y440"
          id="btn-read-more"
        >
          <strong>
            Read more
          </strong>
           about this hotel  
          <i
            class="fas fa-angle-right"
          />
        </button>
      </div>
      <div
        class="readMoreContent_1aeiymr-o_O-isHidden_iq1cgs"
      >
        <h3
          class="h3_1ns9ks7"
        >
          Overview
        </h3>
        <p
          class="p_fhs6vg"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
      </div>
    </section>
    <section>
      <div
        class="card_pmicl5"
      >
        <img
          alt="Image of the hotel"
          class="img_bm229d"
          src="http://localhost:9999/hotel-image-1.png"
        />
        <div
          class="contentContainer_1ydb5eu"
        >
          <h1>
            Really Fancy Hotel
          </h1>
          <h2
            class="h2_1tkgvlr"
          >
            Costa Adeje, Tenerife
          </h2>
          <div>
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
          </div>
          <ul
            class="ul_tx6wom"
          >
            <li>
              <span>
                <strong>
                  2
                </strong>
                 adults
              </span>
            </li>
            <li>
              <strong>
                3rd July 2019
              </strong>
               for 
              <strong>
                7 days
              </strong>
            </li>
            <li>
              departing from 
              <strong>
                East Midlands
              </strong>
            </li>
          </ul>
          <button
            class="callToAction_55n7hg"
          >
            <span
              class="actionText_131r5dh"
            >
              Book now
            </span>
            <span
              class="actionAmount_qtg89e"
            >
              £2,136.50
            </span>
          </button>
        </div>
        <button
          class="readMore_j9y440"
          id="btn-read-more"
        >
          <strong>
            Read more
          </strong>
           about this hotel  
          <i
            class="fas fa-angle-right"
          />
        </button>
      </div>
      <div
        class="readMoreContent_1aeiymr-o_O-isHidden_iq1cgs"
      >
        <h3
          class="h3_1ns9ks7"
        >
          Overview
        </h3>
        <p
          class="p_fhs6vg"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
      </div>
    </section>
    <section>
      <div
        class="card_pmicl5"
      >
        <img
          alt="Image of the hotel"
          class="img_bm229d"
          src="http://localhost:9999/hotel-image-1.png"
        />
        <div
          class="contentContainer_1ydb5eu"
        >
          <h1>
            Really Fancy Hotel v2
          </h1>
          <h2
            class="h2_1tkgvlr"
          >
            Costa Adeje, Tenerife
          </h2>
          <div>
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
          </div>
          <ul
            class="ul_tx6wom"
          >
            <li>
              <span>
                <strong>
                  2
                </strong>
                 adults
              </span>
            </li>
            <li>
              <strong>
                3rd July 2019
              </strong>
               for 
              <strong>
                7 days
              </strong>
            </li>
            <li>
              departing from 
              <strong>
                East Midlands
              </strong>
            </li>
          </ul>
          <button
            class="callToAction_55n7hg"
          >
            <span
              class="actionText_131r5dh"
            >
              Book now
            </span>
            <span
              class="actionAmount_qtg89e"
            >
              £2,136.50
            </span>
          </button>
        </div>
        <button
          class="readMore_j9y440"
          id="btn-read-more"
        >
          <strong>
            Read more
          </strong>
           about this hotel  
          <i
            class="fas fa-angle-right"
          />
        </button>
      </div>
      <div
        class="readMoreContent_1aeiymr-o_O-isHidden_iq1cgs"
      >
        <h3
          class="h3_1ns9ks7"
        >
          Overview
        </h3>
        <p
          class="p_fhs6vg"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
      </div>
    </section>
    <section>
      <div
        class="card_pmicl5"
      >
        <img
          alt="Image of the hotel"
          class="img_bm229d"
          src="http://localhost:9999/hotel-image-3.png"
        />
        <div
          class="contentContainer_1ydb5eu"
        >
          <h1>
            Swank-ville
          </h1>
          <h2
            class="h2_1tkgvlr"
          >
            Costa Adeje, Tenerife
          </h2>
          <div>
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
          </div>
          <ul
            class="ul_tx6wom"
          >
            <li>
              <span>
                <strong>
                  1
                </strong>
                 adult
              </span>
            </li>
            <li>
              <strong>
                3rd July 2019
              </strong>
               for 
              <strong>
                7 days
              </strong>
            </li>
            <li>
              departing from 
              <strong>
                East Midlands
              </strong>
            </li>
          </ul>
          <button
            class="callToAction_55n7hg"
          >
            <span
              class="actionText_131r5dh"
            >
              Book now
            </span>
            <span
              class="actionAmount_qtg89e"
            >
              £5,136.50
            </span>
          </button>
        </div>
        <button
          class="readMore_j9y440"
          id="btn-read-more"
        >
          <strong>
            Read more
          </strong>
           about this hotel  
          <i
            class="fas fa-angle-right"
          />
        </button>
      </div>
      <div
        class="readMoreContent_1aeiymr-o_O-isHidden_iq1cgs"
      >
        <h3
          class="h3_1ns9ks7"
        >
          Overview
        </h3>
        <p
          class="p_fhs6vg"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
      </div>
    </section>
    <section>
      <div
        class="card_pmicl5"
      >
        <img
          alt="Image of the hotel"
          class="img_bm229d"
          src="http://localhost:9999/hotel-image-2.png"
        />
        <div
          class="contentContainer_1ydb5eu"
        >
          <h1>
            Wow - so fancy
          </h1>
          <h2
            class="h2_1tkgvlr"
          >
            Costa Adeje, Tenerife
          </h2>
          <div>
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
            <i
              class="starIcon_17x7h34 fas fa-star"
            />
          </div>
          <ul
            class="ul_tx6wom"
          >
            <li>
              <span>
                <strong>
                  2
                </strong>
                 adults & 
                <strong>
                  1
                </strong>
                 child
              </span>
            </li>
            <li>
              <strong>
                3rd July 2019
              </strong>
               for 
              <strong>
                7 days
              </strong>
            </li>
            <li>
              departing from 
              <strong>
                East Midlands
              </strong>
            </li>
          </ul>
          <button
            class="callToAction_55n7hg"
          >
            <span
              class="actionText_131r5dh"
            >
              Book now
            </span>
            <span
              class="actionAmount_qtg89e"
            >
              £1,136.50
            </span>
          </button>
        </div>
        <button
          class="readMore_j9y440"
          id="btn-read-more"
        >
          <strong>
            Read more
          </strong>
           about this hotel  
          <i
            class="fas fa-angle-right"
          />
        </button>
      </div>
      <div
        class="readMoreContent_1aeiymr-o_O-isHidden_iq1cgs"
      >
        <h3
          class="h3_1ns9ks7"
        >
          Overview
        </h3>
        <p
          class="p_fhs6vg"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
      </div>
    </section>
  </article>
</div>
`;
